Add main menu

woubuc opened the pull request:
I added a basic main menu and an empty settings menu panel. You can now click "Play" to start the game, and while in-game press escape to go back to the main menu. It's almost like a real game! 😄 

![image](https://user-images.githubusercontent.com/1015540/107953369-5a44bb80-6f9b-11eb-9431-aa290ca1494b.png)

**Bug/workaround**
I prevented the cameras from despawning (with FIXME comments) because of an issue in Bevy (bevyengine/bevy#1452) where the cameras seemingly don't get removed from the `ActiveCameras`, causing the game to panic. So if you go back and forth between the menus and the game several times, you may end up with a bunch of unused cameras somewhere(?). But everything still seems to work even if we don't despawn the cameras so I think this is fine at this point.

woubuc committed:
Add editorconfig

woubuc committed:
Merge remote-tracking branch 'upstream/master'

woubuc committed:
Add main menu & settings menu

woubuc committed:
Style tweaks

woubuc committed:
Workaround for camera despawn bug

woubuc committed:
Change TODO into FIXME

inodentry commented:
I am really unsure about the font license. This is something I want to be discussed with everyone before merging, if using this font.

guimcaballero commented:
My only comment would be to either move font loading into `MenuMaterials` or make a new resource to hold the fonts, instead of having a bunch of `asset_server.load("fonts/sansation/sansation_light.ttf")`.

Everything else looks good!

woubuc commented:
> I am really unsure about the font license. This is something I want to be discussed with everyone before merging, if using this font.

What concerns you about the license?

inodentry commented on src/main.rs:
```diff
@@ -39,13 +43,23 @@ fn main() {
+        // Settings menu
+        .on_state_enter(APPSTATES, AppState::SettingsMenu, menu::settings::setup.system())
+        .on_state_update(APPSTATES, AppState::SettingsMenu, menu::button_interact.system())
+        .on_state_exit(APPSTATES, AppState::SettingsMenu, despawn_all::<menu::StateCleanup>.system())
```
I think it is better if you don't reuse the same marker type for main menu and settings menu entities.

It will be less prone to breakage if we want to change the flow in the future.

Ratysz replied:
Isn't it a different type, just named the same? They're from different modules.

woubuc replied:
> Isn't it a different type, just named the same? They're from different modules.

It's a different one from the game one, but I did reuse it for the main menu & settings menu. I've now fixed this and used separate marker components for both.

inodentry replied:
Umm, no? `menu::StateCleanup` and `overworld::StateCleanup` are separate types. But it seems to me that both menus are reusing the same `menu::StateCleanup`.

Ratysz replied:
Right, I missed that part. It is indeed fixed now.

woubuc commented:
> My only comment would be to either move font loading into `MenuMaterials` or make a new resource to hold the fonts, instead of having a bunch of `asset_server.load("fonts/sansation/sansation_light.ttf")`.

That's a fair point! I'll rename it the struct to a more general `MenuAssets` and preload all fonts in there as well.

woubuc committed:
Use separate `StateCleanup` marker components for main menu & settings menu

inodentry commented on src/menu/main_menu.rs:
```diff
@@ -0,0 +1,162 @@
+                    text: Text::with_section(
+                        "a bevy community game",
+                        TextStyle {
+                            font: asset_server.load("fonts/sansation/sansation_light.ttf"),
```
It seems brittle to just hardcode paths to assets all over the place like that.

I think you should extend `MenuMaterials` to become a larger `UiConfig` (or something along those lines) to hold all the static things that need to be referenced when building the UIs: styles, font, assets, materials, etc. all together. Load/configure them all at once in a single place at setup time.

It would help the code be less brittle, as there would be a single place for all the UI building stuff. It would also make it more convenient to add more UIs to the game, as you can just reference anything you need from one place.

I think one monolithic resource for this stuff is fine, as this is all stuff that is to be loaded/configured once at startup and then only accessed read-only. We wouldn't win anything by making it more granular.

woubuc replied:
I started from Bevy's button UI example where it only uses a resource for the materials so I just implemented it the same way, but it makes total sense to load all assets centrally and make them available trough in a resource. Fixing this right now! 😄

TheRawMeatball commented on src/menu/main_menu.rs:
```diff
@@ -0,0 +1,162 @@
+    };
+
+    commands
+        .spawn(UiCameraBundle::default())
```
Hmm, we should probably make the UI camera state independent - we'll have _some_ UI in all states, and this avoids a nasty workaround

inodentry replied:
yes agreed, i think we should just spawn the UI camera at app launch (you can even do it in a start up system outside of states), and just keep it alive forever.

There is no harm, we are probably gonna have lots of UIs in every game state anyway.

It will just serve to make sure that it can always be displayed.

woubuc replied:
Sounds good. I've replaced the UI cameras with a single global UI camera.

woubuc committed:
Load all necessary menu assets in a resource instead of ad-hoc during setup systems.

inodentry commented on src/menu/mod.rs:
```diff
@@ -0,0 +1,78 @@
+            Interaction::Clicked => {
+                *material = materials.button_active.clone();
+
+                match action {
```
Having one system handle all button presses isn't gonna scale. We would need to add more and more parameters to this function in the future as we add more buttons to the game.

However, I think you can leave this for now. I have a proposal for how to solve this elegantly, and I can submit that as a subsequent PR after we merge this one.

woubuc replied:
Yup, this isn't intended to be a "final" solution in any way, I just wanted to make the menu buttons work 😄

inodentry replied:
Yep, this is the mindset to have. None of what we are doing here is "final", just have fun and make it work :smile: 

But I also feel obligated to leave notes about important points that we should all be aware of, so that my thoughts are put down in writing.

woubuc committed:
Replace menu cameras with global UI camera

inodentry commented:
> What concerns you about the license?

It is restrictive. It does not allow distributing modified versions, and it does not allow commercial use.

We picked MIT for code and CC-BY-SA 2.0 for assets, because we want the project to be permissively licensed and leave a possibility for commercial use in the future.

guimcaballero commented:
IMO it's fine at the moment, we can always use another font later if we decide we need to make modifications or we want to publish the game commercially.

inodentry commented:
> IMO it's fine at the moment, we can always use another font later if we decide we need to make modifications or we want to publish the game commercially.

I am also concerned about the project becoming too much of a mix of different licenses. Particularly non-standard arbitrary licenses like that with questionable legal status. It could become a mess.

inodentry commented:
OK, for the license thing, I think allowing a wider mix of licenses into the project would be fine, but we need to communicate it clearly.

For now, can you just update the README to not say that all assets are CC-BY-SA, and create a separate file (maybe something like `LICENSES.ASSETS`) where we can keep a list of licenses we have in the project, along with what they cover.

woubuc committed:
Replace font with OFL-licensed font

woubuc commented:
I've changed the font to a properly OFL-licensed font (no more custom license) and updated the readme to reflect the use of this license. I don't think adding a separate license file is necessary at this point if we standardise on using OFL for fonts.

woubuc committed:
Cargo fmt
